### PR TITLE
latest Arc fixes

### LINKED ISSUE #8092: tests/kernel/fatal crashes on Quark SE / ARC
commit 43553da9b2bb9e537185695899fb5184c3b17ebe

different crash examples:

```
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS v1.12.0-rc2-45-g43553da9b2 (delayed boot 1000ms) *****
Running test suite fatal
===================================================================
starting test - test_fatal
Exception vector: 0x6, cause code: 0x1, parameter 0x2
Address 0xa80011e4
***** Stack Check Fail! *****
Current thread ID = 0xa8000540
Caught system error -- reason 2
ASSERTION FAIL [!(thread->base.user_options & (1 << 0))] @ /home/inaky/z/kernel.git/kernel/thread_abort.c:35:
	essential thread aborted
```
another:

```
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS 1.12.0-rc2 (delayed boot 1000ms) *****
Running test suite fatal
===================================================================
starting test - test_fatal
test alt thread 1: generic CPU exception
Exception vector: 0x8, cause code: 0x0, parameter 0x0
Address 0x400007cc
Current thread ID = 0xa8000400
Faulting instruction address = 0x400007cc
Caught system error -- reason 0
test alt thread 2: initiate kernel oops
***** Kernel OOPS! *****
Current thread ID = 0xa8000400
Caught system error -- reason 4
test alt thread 3: initiate kernel panic
***** Kernel Panic! *****
Current thread ID = 0xa8000400
Caught system error -- reason 5
test stack overflow - timer irq
```

### PR DESCRIPTION
**Pull request by vonhust**
This PR tries to fix
#8092 
  commit c058be6
  commit ca651b3
#8032 
  commit 5d58f0f
#6515
  commit f8509cc

### CODE REVIEW

**Review by nashif — APPROVED** [review]
I confirm both high priority bugs are fixed with this.

File: arch/arc/core/fault_s.S [hunk 1]
```diff
@@ -36,6 +36,15 @@ GTEXT(__ev_maligned)
 
 GDATA(exc_nest_count)
 
+	.balign 4
+SECTION_VAR(BSS, saved_value)
+	.word 0
+
+#define EXCEPTION_STACK_SIZE 256
```
Comment (andrewboie) [inline arch/arc/core/fault_s.S]: instead of doing this, why not use the interrupt stack for exceptions like all our other arches?

**Review by tyrantlizard77 — COMMENTED** [review]
generally looks good but I don't understand why we don't handle exceptions on the IRQ stack, all other arches do that

**Comment by ruuddw** [comment]
@andrewboie

> instead of doing this, why not use the interrupt stack for exceptions like all our other arches?

 The interrupt stack was used in the original code, but only after initial exception context was saved to the current stack. The problem being fixed here is the following (#8092)
1. the test code creates a stack overflow; sp gets out of bounds because of a large array on the stack, and then a printf call is done
2. exception handler is triggered (hardware stack checking), and creates an exception stack frame … using the incorrect sp
3. _kernel struct gets corrupted due to this, and later in the exception handler the switch to _interrupt_stack (fetched from the _kernel struct) fails since it was overwritten.

I don't think the interrupt stack can be trusted in case of stack overflow exceptions. Hence this solution with a fail safe dedicated exception stack.

**Comment by andrewboie** [comment]
I still don't think we should do this. That's a lot of space to reserve for what should not happen in practice.

> _kernel struct gets corrupted due to this, and later in the exception handler the switch to _interrupt_stack (fetched from the _kernel struct) fails since it was overwritten.

This is the real problem. There's no need that I can see to have to store the pointer to the IRQ stack in _kernel. It gets initialized once to K_THREAD_STACK_BUFFER(_interrupt_stack) + CONFIG_ISR_STACK_SIZE and is never touched again. It's essentially a read-only value. It doesn't need to be in RAM.

Instead I'd like to see if we can  remove the irq_stack members from _kernel (for both UP and SMP) and use ROM for it:

* initialize a const read-only array at buiild time which contains this information and stored in ROM/text region and have the exception handler read that
* or if that turns out to be impossible, just have the fatal exception handler use the address of the _interrupt_stack symbol, add CONFIG_ISR_STACK_SIZE to it, and set the stack pointer to that.

Can you drop the exception stack patch from this PR? I have time to work on what I have described above.

**Review by tyrantlizard77 — DISMISSED** [review]
I've opened #8271 to track putting the irq stack pointers in ROM

**Comment by ruuddw** [comment]
Yes, while analysing the stack issues I wondered why the interrupt stack was obtained from the kernel struct since it is constant. But, that is not the issue. We support nested (different priority) interrupts, and the interrupts can trigger an exception as well. Setting sp to the initial interrupt stack value as the first action in the exception handler interferes with that. It would probably be ok if exceptions don't have to be recoverable, that would greatly simplify things.

**Comment by andrewboie** [comment]
> We support nested (different priority) interrupts, and the interrupts can trigger an exception as well. Setting sp to the initial interrupt stack value as the first action in the exception handler interferes with that. 

sp only gets set to _kernel interrupt stack if we are not already on the interrupt stack (nested count = 0)
If we are already in an IRQ or exception, sp should not be modified. This is true for all arches AFAIK, if ARC is unconditionally setting sp on exception entry then that is a bug.

**Comment by ruuddw** [comment]
Exceptions (on ARC) can't be nested, an exception during exception will trigger a non-recoverable double fault. So unconditional switching to the dedicate exception stack is ok. The nesting is required for the interrupt stack.
The alternative to the exception stack could indeed be a check if the sp is in range of the _kernel interrupt stack and switch to it if not. That would result in a safe sp for the exception handling. Need to give that a bit more thought...

**Comment by nashif** [comment]
We need a resolution here, this is blocking the 1.12 release.

**Comment by andrewboie** [comment]
> We need a resolution here, this is blocking the 1.12 release.

I don't want to block the release, but if we take the current approach of an exception stack I feel it's mandatory to work on this some more in 1.13 so we don't lose that memory.

**Review by MaureenHelm — APPROVED** [review]

**Comment by MaureenHelm** [comment]
Fixes #8092
Fixes #8032
Fixes #6515

**Pull request merged**